clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -162,6 +163,13 @@ public void connect(String id, InetSocketAddress address, int sendBufferSize, in
         KafkaChannel channel = channelBuilder.buildChannel(id, key, maxReceiveSize);
         key.attach(channel);
         this.channels.put(id, channel);
+
+        if (connected) {
```

Comment:
Agreed. It's definitely hacky but it appeared that a lot of code depended on events being handled in Selector.poll() and I wanted to minimize the scope of the change. I initially did as you suggested but was not successful. I didn't pursue that line for very long since it would violate what I perceived as the contract with Selector.poll().